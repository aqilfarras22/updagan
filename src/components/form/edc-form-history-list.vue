<template>
  <table-component
    showEmpty
    noGutters
    :items="history"
    :fields="formHistoryFields"
  >
    <!-- date -->
    <template slot="createdAt" slot-scope="data">
      {{ new Date(data.row.item.createdAt) | formatUTCDate }}
    </template>
    <!-- action -->
    <template slot="eventType" slot-scope="data">
      {{ data.row.item.primaryReason && data.row.item.primaryReason!='' ? data.row.item.primaryReason : (verificationStatusText[data.row.item.eventType] || approvalStatusText[data.row.item.eventType] || historyStatusText[data.row.item.eventType]) }}
    </template>
    <!-- view -->
    <template slot="view" slot-scope="data">
      <a v-if="historyStatus.MODIFIED === data.row.item.eventType" @click.prevent="view(data.row.item)" href="">View</a>
    </template>
  </table-component>
</template>

<script>
import TableComponent from '@/components/common/table';
import {
  approvalStatusText,
  verificationStatusText,
  historyStatusText,
  historyStatus,
} from '@/constants/form';

export default {
  name: 'form-history-list',
  components: {
    TableComponent,
  },
  props: {
    history: {
      type: Array,
      default: null,
    },
  },
  data() {
    return {
      approvalStatusText,
      verificationStatusText,
      historyStatusText,
      historyStatus,
      formHistoryFields: [
        {
          key: 'createdAt',
          label: 'Date',
          sortable: true,
        },
        {
          key: 'eventType',
          label: 'Action / Changes',
          sortable: true,
        },
        {
          key: 'createdBy',
          label: 'By',
          sortable: true,
        },
        {
          key: 'view',
          label: ' ',
        },
      ],
    };
  },
  methods: {
    view(item) {
      this.$emit('view', item);
    },
  },
};
</script>
