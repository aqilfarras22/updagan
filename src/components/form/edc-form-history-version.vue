<template>
  <div>
    <b-row class="border-bottom header-row">
      <b-col><a @click.prevent="back" class="backButton" href=""><i class="fa fa-caret-left fa-lg"></i>&nbsp; Back to history</a></b-col>
      <b-col class="text-right"><i class="square"></i>&nbsp; <span class="info">Edited fields</span></b-col>
    </b-row>
    <div>
      <b-card-group v-if="!edcFormVersionData.isUnSchedule" class="pb-15">
        <b-card no-body class="border-0" :header="edcFormVersionData.name">
          <div v-if="edcFormVersionData.primaryReason" class="reason">
            Reason for change: {{ edcFormVersionData.primaryReason }}
          </div>
          <SubmittedFormPreview v-if="edcFormVersionData.fields" :isHistory="true" :formData="edcFormVersionData.fields" />
        </b-card>
      </b-card-group>
    </div>
    <div v-if="edcFormVersionData.isUnSchedule && edcFormVersionData.fields">
      <b-card-group v-for="(unscheduleAnswers, index) in edcFormVersionData.fields" :key="index" :class="{'border-top': index!=0 ,'pb-15' : index+1 === edcFormVersionData.fields.length, 'pb-5' : index+1 !== edcFormVersionData.fields.length}">
        <b-card no-body class="border-0" :header="unscheduleAnswers.formName">
          <div v-if="edcFormVersionData.primaryReason" class="reason">
            Reason for change: {{ edcFormVersionData.primaryReason }}
          </div>
          <SubmittedFormPreview v-if="unscheduleAnswers.formData" :isHistory="true" :formData="unscheduleAnswers.formData"/>
        </b-card>
      </b-card-group>
    </div>
  </div>
</template>

<script>
import SubmittedFormPreview from '@/components/edc-form/submitted-form-preview';

export default {
  name: 'form-history-version',
  components: { SubmittedFormPreview },
  props: {
    edcFormVersionData: Object,
  },
  methods: {
    back() {
      this.$emit('back');
    },
  },
};
</script>

<style scoped>
.header-row {
  padding: 0.9rem 0 0.65rem;
}
.square {
  display: inline-block;
  width: 16px;
  height: 16px;
  border-radius: 1px;
  margin-bottom: -2px;
  background-color: #ffa200;
}
.info {
  font-size: 12px;
  display: inline-block;
}
.reason {
  font-family: "RobotoBold" !important;
  color: #666669;
  padding-left: 30px;
  padding-top: 19px;
}
</style>
