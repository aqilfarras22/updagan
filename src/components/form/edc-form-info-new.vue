<template>
  <div class="edc-form-bar">
	<div class="row">
		<div class="col-md-2">
			<div class="heading">Form/Packet Name</div>
			<div class="value">{{ clearFormName() }}</div>
		</div>
		<div class="col-md-2">
			<div class="heading">Verification Status</div>
			<div class="value" :class="edcFormInfo.verificationStatus">{{ verificationStatusText[edcFormInfo.verificationStatus] }}</div>
			<div v-if="verified && edcFormInfo.verifiedOn" class="date value">
				({{ new Date(edcFormInfo.verifiedOn) | formatGMTDate }})
			</div>
		</div>
		<div class="col-md-2">
			<div class="heading">Approval Status</div>
			<div class="value" :class="edcFormInfo.approvalStatus">{{ approvalStatusText[edcFormInfo.approvalStatus] }}</div>
			<span v-if="approved && edcFormInfo.approvedOn" class="date value">
			({{ new Date(edcFormInfo.approvedOn) | formatGMTDate }})
			</span>
		</div>
	</div>
  </div>
</template>

<script>
import { AclRule } from 'vue-acl';
import { mapGetters } from 'vuex';
import {
  approvalStatusText,
  verificationStatusText,
  verificationStatus,
  approvalStatus,
} from '@/constants/form';

export default {
  name: 'form-info',
  components: {},
  props: {
    edcFormInfo: {
		type: Object,
		default: {}
	}
  },
  data() {
    return {
      approvalStatusText,
	  verificationStatusText,
	  verificationStatus
	};
  },
  computed: {
    ...mapGetters(['user']),
    userAccess() {
      return new AclRule(this.user.userRole).generate();
    }
  },
  methods: {
	clearFormName() {
    	return this.edcFormInfo.formName ? this.edcFormInfo.formName.replace(/_/g, ' ') : '';
	},
	verified() {
      	return this.edcFormInfo.verificationStatus === verificationStatus.VERIFIED;
    },
    approved() {
      	return this.edcFormInfo.approvalStatus === approvalStatus.APPROVED;
    }
  }
};
</script>

<style scoped lang="scss">
.edc-form-bar {
	height: 60px;
  background-color: #fff;
  padding: 10px 0px;
  -webkit-box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.06);
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.06);
  margin-top: 15px;
  height: 75px;
  .heading {
    font-weight: 600;
  }
}
.date {
  font-size: 12px;
  line-height: 1.67;
}
.telehealth-call {
  .heading {
    font-size: 10px;
  }
  .value {
    font-size: 10px;
  }
}
</style>
