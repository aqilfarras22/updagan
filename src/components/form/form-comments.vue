<template>
  <div>
    <b-card-group class="border-top">
      <b-card no-body class="border-0" :header="title">
        <b-list-group v-if="comments" flush>
          <b-list-group-item class="item border-0" v-for="comment in comments" :key="comment.id">
            <b-row>
              <div class="col-5">
                <div class="name-class pb-5">
                  <span v-if="comment.author.name.firstName">{{ comment.author.name.firstName }}</span>
                  <span v-if="comment.author.name.middleName">{{ comment.author.name.middleName }}</span>
                  <span v-if="comment.author.name.lastName">{{ comment.author.name.lastName }}</span>
                </div>
                <div>{{ new Date(comment.postedAt * 1000 ) | formatGMTDate }}</div>
              </div>
              <div class="col-7 div_style">{{ comment.message }}</div>
            </b-row>
          </b-list-group-item>
        </b-list-group>
      </b-card>
    </b-card-group>
  </div>
</template>

<script>
import ButtonCmp from '@/components/common/button';

export default {
  name: 'form-fields-list',
  components: { ButtonCmp },
  props: {
    comments: Array,
  },
  data() {
    return {
      title: 'Comments',
    };
  },
  methods: {},
};
</script>

<style scoped>
.item {
  font-size: 0.9rem;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.43;
  text-align: left;
  color: #666669;
  padding-left: 30px;
}
.name-class {
  font-weight: 700;
}
.div_style {
  border-left: 1px solid #d4d9e2; 
}
</style>
