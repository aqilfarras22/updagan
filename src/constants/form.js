export const fieldTypes = {
  POLAR_QUESTION: 'POLAR_QUESTION',
  DROPDOWN: 'DROPDOWN',
  DATEPICKER: 'DATEPICKER',
  INPUT: 'INPUT',
  LABEL: 'LABEL',
};

export const formFieldQueryStatus = {
  OPEN: 'OPEN',
  CLOSE: 'CLOSE',
};

export const historyStatus = {
  MODIFIED: 'MODIFIED',
};

export const historyStatusText = {
  MODIFIED: 'Original Submission',
};

// approve
export const approvalStatus = {
  APPROVED: 'APPROVED',
  NOT_APPROVED: 'NOT_APPROVED',
};
export const approvalStatusText = {
  APPROVED: 'Approved',
  NOT_APPROVED: 'Unapproved',
};

// verification
export const verificationStatus = {
  VERIFIED: 'VERIFIED',
  NOT_VERIFIED: 'NOT_VERIFIED',
};
export const verificationStatusText = {
  VERIFIED: 'Verified',
  NOT_VERIFIED: 'Unverified',
};
