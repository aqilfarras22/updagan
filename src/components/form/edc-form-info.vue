<template>
  <div v-if="formId">
    <dl v-if="formName" ref="formName">
      <dt>Form/Packet Name</dt>
      <dd>{{ clearFormName }}</dd>
    </dl>
    <dl ref="verificationStatus">
      <dt>Verification Status</dt>
      <dd>
        <div :class="verificationStatus">{{ verificationStatusText[verificationStatus] }}</div>
        <div v-if="verified && verifiedOn" class="date">
          ({{ new Date(verifiedOn) | formatGMTDate }})
        </div>
      </dd>
    </dl>
    <dl ref="approvalStatus">
      <dt>Approval Status</dt>
      <dd>
        <div :class="approvalStatus">{{ approvalStatusText[approvalStatus] }}</div>
        <span v-if="approved && approvedOn" class="date">
          ({{ new Date(approvedOn) | formatGMTDate }})
        </span>
      </dd>
    </dl>
    <div class="actions">
      <button-cmp v-if="canPiAccess || canSubPiAccess || canScAccess || canHomeHealthAccess" class="form-action-button" size="lg" variant="primary" @click="openApprovalModal">{{ approvalBtnText }}</button-cmp>
      <button-cmp v-if="canCraAccess" class="form-action-button" size="lg" variant="primary" @click="openVerificationModal">{{ verifyBtnText }}</button-cmp>
    </div>
    <formApprovalModal
      v-if="canPiAccess || canSubPiAccess || canScAccess || canHomeHealthAccess"
      ref="approvalModal"
      callApi
      :formId="formId"
      :approved="approved"
      @confirmed="approvalConfirmed"
    />
    <formVerificationModal
      v-if="canCraAccess"
      ref="verificationModal"
      callApi
      :formId="formId"
      :verified="verified"
      @confirmed="verificationConfirmed"
    />
  </div>
</template>

<script>
import { AclRule } from 'vue-acl';
import { mapGetters } from 'vuex';
import formVerificationModal from '@/components/modals/edc-form-verification-modal';
import formApprovalModal from '@/components/modals/edc-form-approval-modal';
import ButtonCmp from '@/components/common/button';

import { PI, CRA, SUB_PI, STUDY_COORDINATOR, HOME_HEALTH } from '@/constants/authRoles';
import {
  approvalStatusText,
  verificationStatusText,
  verificationStatus,
  approvalStatus,
} from '@/constants/form';

export default {
  name: 'form-info',
  components: {
    formVerificationModal,
    formApprovalModal,
    ButtonCmp,
  },
  props: {
    formId: String,
    formName: String,
    approvalStatus: String,
    verificationStatus: String,
    verifiedOn: [String, Number],
    approvedOn: [String, Number],
  },
  data() {
    return {
      approvalStatusText,
      verificationStatusText,
    };
  },
  computed: {
    ...mapGetters(['user']),
    userAccess() {
      return new AclRule(this.user.userRole).generate();
    },
    canPiAccess() {
      this.$acl.change(`${PI}`);
      return this.$acl.check('userAccess');
    },
    canSubPiAccess() {
      this.$acl.change(`${SUB_PI}`);
      return this.$acl.check('userAccess');
    },
    canScAccess() {
      this.$acl.change(`${STUDY_COORDINATOR}`);
      return this.$acl.check('userAccess');
    },
    canHomeHealthAccess() {
      this.$acl.change(`${HOME_HEALTH}`);
      return this.$acl.check('userAccess');
    },
    canCraAccess() {
      this.$acl.change(`${CRA}`);
      return this.$acl.check('userAccess');
    },
    clearFormName() {
      return this.formName ? this.formName.replace(/_/g, ' ') : '';
    },
    verified() {
      return this.verificationStatus === verificationStatus.VERIFIED;
    },
    approved() {
      return this.approvalStatus === approvalStatus.APPROVED;
    },
    verifyBtnText() {
      return this.verified ? 'Unverify' : 'Verify';
    },
    approvalBtnText() {
      return this.approved ? 'Unapprove' : 'Approve';
    },
  },
  methods: {
    openVerificationModal() {
      this.$refs.verificationModal.show();
    },
    openApprovalModal() {
      this.$refs.approvalModal.show();
    },
    verificationConfirmed({ verified }) {
      const status = verified ? verificationStatus.VERIFIED : verificationStatus.NOT_VERIFIED;
      this.$emit('verificationConfirmed', status);
    },
    approvalConfirmed({ approved }) {
      const status = approved ? approvalStatus.APPROVED : approvalStatus.NOT_APPROVED;
      this.$emit('approvalConfirmed', status);
    },
  },
};
</script>

<style scoped>
.APPROVED,
.VERIFIED {
  color: #48bf7a !important;
}
.NOT_VERIFIED,
.NOT_APPROVED {
  color: #f16559 !important;
}
.form-action-button {
  width: 100%;
  margin-top: 1px;
  margin-bottom: 1px;
}
dt {
  font-size: 10px;
  line-height: 1.6;
  text-transform: uppercase;
  color: #adaeb1;
}
dd {
  color: #666669;
  font-size: 14px;
  font-weight: 500;
}
.date {
  font-size: 12px;
  line-height: 1.67;
}
</style>
