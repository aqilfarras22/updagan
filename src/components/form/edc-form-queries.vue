<template>
  <b-tabs class="" v-model="tabIndex">
    <b-tab :title="`Open Queries: ${openQueries.length}`" :title-link-class="queriesOpen()">
      <FormQueriesCommentForm v-if="value" @cancel="cancelQuery" @submit="createQuery"/>
      <div v-else>
        <div v-if="openQueryId">
          <FormQueriesQueryDetails
            :userNames="userNames"
            :comments="comments"
            :query="openQuery"
            @back="hideOpenQuery"
            @onQueryCommentCreate="onQueryCommentCreate"
            @onQueryClosed="onQueryClosed"
          />
        </div>
        <div v-else>
          <FormQueriesHeader v-if="filter" :title="`Filtered by field ID ${shortFilterId}`" @back="resetFilter" backTitle="Clear"/>
          <b-row v-if="!openQueries.length" class="py-2">
            <b-col><p>No queries created yet.<br>Select a field to create a query.</p></b-col>
          </b-row>
          <FormQueriesList
            v-else
            :queries="openQueries"
            @view="viewOpenQuery"
          />
        </div>
      </div>
    </b-tab>
    <b-tab :title="`Closed Queries: ${closedQueries.length}`">
      <div v-if="closedQuery">
        <FormQueriesQueryDetails
          closed
          :comments="comments"
          :userNames="userNames"
          :query="closedQuery"
          @back="hideClosedQuery"
          @onQueryReopened="onQueryReopened"
        />
      </div>
      <div v-else>
        <FormQueriesHeader v-if="filter" :title="`Filtered by field ID ${filter}`" @back="resetFilter" backTitle="Clear"/>
        <b-row v-if="!closedQueries.length">
          <b-col class="py-2"><p>No queries closed yet.</p></b-col>
        </b-row>
        <FormQueriesList
          v-else
          :queries="closedQueries"
          @view="viewClosedQuery"
        />
      </div>
    </b-tab>
  </b-tabs>
</template>

<script>
import sortBy from 'lodash/sortBy';
import FormQueriesList from '@/components/form/edc-form-queries/form-queries-list';
import FormQueriesQueryDetails from '@/components/form/edc-form-queries/form-queries-query-details';
import FormQueriesCommentForm from '@/components/form/edc-form-queries/form-queries-comment-form';
import FormQueriesHeader from '@/components/form/form-queries-header';

import { formFieldQueryStatus } from '@/constants/form';

export default {
  name: 'edc-form-queries',
  components: {
    FormQueriesList,
    FormQueriesQueryDetails,
    FormQueriesCommentForm,
    FormQueriesHeader,
  },
  props: {
    value: String,
    queries: Array,
    filter: String,
    formId: String,
    studyId: String,
    comments: Array,
  },
  data() {
    return {
      tabIndex: 0,
      openQueryId: null,
      closedQueryId: null,
      userNames: {},
    };
  },
  async mounted() {
    const type="studyteam"
    const associated = "true"
    const studyTeamData = await this.$api.getStudyUsersFact(this.$route.params.studyId, type, associated);
    const siteTeamData = await this.$api.getUserNames(this.$route.params.studyId);
    studyTeamData.body.forEach(team => {
      this.userNames[team.id] = `${team.firstName} ${team.lastName}`;
    });
    siteTeamData.body.forEach(team => {
      this.userNames[team.id] = `${team.firstName} ${team.lastName}`;
    });
  },
  computed: {
    shortFilterId() {
      if (this.filter) {
        // return this.filter.slice(0, 5).toUpperCase();
        return this.filter;
      }
      return '';
    },
    filteredQueries() {
      let filtered = this.queries;
      if (this.filter) {
        filtered = this.queries.filter(q => q.fieldId === this.filter);
        this.$forceUpdate();
      }
      return sortBy(filtered, ['openedOn']).reverse();
    },
    openQueries() {
      return this.filteredQueries.filter(q => q.status === formFieldQueryStatus.OPEN);
    },
    closedQueries() {
      return this.filteredQueries.filter(q => q.status !== formFieldQueryStatus.OPEN);
    },
    openQuery() {
      return this.queries.find(q => q.queryId === this.openQueryId);
    },
    closedQuery() {
      return this.queries.find(q => q.queryId === this.closedQueryId);
    },
  },
  watch: {
    value(val) {
      if (val) {
        this.openQueryId = null;
        this.closedQueryId = null;
        this.tabIndex = 0;
      }
    },
    filter(val) {
      if (!this.value) {
        if (val) {
          this.shortFilterId = val;
          this.openQueryId = null;
          this.closedQueryId = null;
          this.tabIndex = 0;
        }
      }
    },
  },
  methods: {
    queriesOpen() {
      return this.filteredQueries.filter(q => q.status === formFieldQueryStatus.OPEN).length > 0 ? 'text-warning' : '';
    },
    viewOpenQuery(queryId) {
      this.openQueryId = queryId;
      const field = this.openQueries.find(q => q.queryId === queryId);
      this.$emit('updateComment', queryId, field.fieldId);
    },
    hideOpenQuery() {
      this.$emit('updateQueries');
      this.openQueryId = null;
      this.$emit('resetFilter');
    },
    viewClosedQuery(queryId) {
      this.closedQueryId = queryId;
      const field = this.closedQueries.find(q => q.queryId === queryId);
      this.$emit('updateComment', queryId, field.fieldId);
    },
    hideClosedQuery() {
      this.closedQueryId = null;
      this.$emit('resetFilter');
    },
    createQuery(commentData) {
      const newQuery = {
        fieldId: this.value,
        formId: this.formId,
        comment: commentData,
      };
      this.$emit('createQuery', newQuery);
      this.$emit('input', null);
    },
    onQueryCommentCreate(commentData) {
      this.$emit('createComment', commentData);
    },
    cancelQuery() {
      this.$emit('input', null);
      this.$emit('resetFilter');
    },
    resetFilter() {
      this.$emit('resetFilter');
      window.scrollTo(0, 0);
    },
    onQueryClosed(query) {
      this.openQueryId = null;
      this.$emit('onQueryClosed', query);
    },
    onQueryReopened(query) {
      this.closedQueryId = null;
      this.$emit('onQueryReopened', query);
    },
  },
};
</script>
<style>
a.text-warning:hover, a.text-warning:focus, .text-warning {
  color: #f5a623 !important;
}
</style>
