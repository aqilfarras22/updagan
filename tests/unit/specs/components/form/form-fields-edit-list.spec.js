import BootstrapVue from 'bootstrap-vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import { fieldTypes } from '@/constants/form';

import formInfoComponent from '@/components/form/form-fields-edit-list';

const { it, describe, expect } = global;

const inputQuestion = {
  fieldId: '121input',
  fragmentType: fieldTypes.INPUT,
  question: 'question',
  response: ['answer'],
};
const polarQuestion = {
  fieldId: '123polar',
  fragmentType: fieldTypes.POLAR_QUESTION,
  question: 'question',
  answers: [{ name: 'Yes' }, { name: 'No' }],
  response: ['Yes'],
};
const dropdownQuestion = {
  fieldId: '421dropdown',
  fragmentType: fieldTypes.DROPDOWN,
  multipleChoice: false,
  question: 'question',
  response: ['first'],
  answers: ['first', 'second', 'third'],
};
const dateQuestion = {
  fieldId: '411date',
  fragmentType: fieldTypes.DATEPICKER,
  question: 'question',
  response: ['1519138199'],
};
const section = {
  name: 'section',
  fields: [inputQuestion, polarQuestion, dropdownQuestion, dateQuestion],
};
const changeReason = 'reason for change';

const itemSelector = '.item';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

const createMountedWrapper = (options = {}) => shallowMount(formInfoComponent, {
  localVue,
  propsData: { sections: [section] },
  ...options,
  stub: {
    FormChangeReasonModal: {
      name: 'FormChangeReasonModal',
      render: h => h('b', 'FormChangeReasonModalCmp'),
    },
  },
});

describe('components/form/form-fields-list.vue', () => {
  it('should create instance', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it('should show correct number of questions', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.findAll(itemSelector).length).toBe(4);
  });

  it('should emit "save" event on submit with unchanged fields', async () => {
    const wrapper = createMountedWrapper();
    wrapper.vm.submit(changeReason);

    await wrapper.vm.$nextTick();
    const actualSaveData = wrapper.emitted('save')[0][0];

    expect(actualSaveData.primaryReason).toBe(changeReason);
    expect(actualSaveData.sections).toEqual([
      {
        name: section.name,
        fields: [inputQuestion, polarQuestion, dropdownQuestion, dateQuestion].map(field => ({
          ...field,
          changed: false,
          oldResponse: field.response,
        })),
      },
    ]);
  });

  it('should emit "save" event on submit with changed fields', async () => {
    const wrapper = createMountedWrapper();

    wrapper.vm.sectionsModel[0].fields[0].response = 'new_input';
    wrapper.vm.submit(changeReason);

    await wrapper.vm.$nextTick();
    const actualSaveData = wrapper.emitted('save')[0][0];

    expect(actualSaveData.primaryReason).toBe(changeReason);
    expect(actualSaveData.sections).toEqual([
      {
        name: section.name,
        fields: [
          {
            ...inputQuestion,
            response: ['new_input'],
            changed: true,
            oldResponse: inputQuestion.response,
          },
          ...[polarQuestion, dropdownQuestion, dateQuestion].map(field => ({
            ...field,
            changed: false,
            oldResponse: field.response,
          })),
        ],
      },
    ]);
  });

  it('has the expected html structure', () => {
    const wrapper = shallowMount(formInfoComponent, {
      localVue,
      propsData: { sections: [section] },
    });
    expect(wrapper.vm.$el).toMatchSnapshot();
  });
});
