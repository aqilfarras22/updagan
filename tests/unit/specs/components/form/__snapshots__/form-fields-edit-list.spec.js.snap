// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/form/form-fields-list.vue has the expected html structure 1`] = `
<b-form-stub>
  <b-card-group-stub
    deck="true"
    tag="div"
  >
    <b-card-stub
      bodytag="div"
      class="border-0"
      footertag="div"
      header="section"
      headertag="div"
      nobody="true"
      subtitletag="h6"
      tag="div"
      titletag="h4"
    >
      <b-list-group-stub
        flush="true"
        tag="div"
      >
        <b-list-group-item-stub
          activeclass="active"
          class="item border-0"
          event="click"
          exactactiveclass="active"
          routertag="a"
          tag="div"
          target="_self"
        >
          <label>
            <strong>
              1. question
            </strong>
          </label>
           
          <div>
            <b-form-input-stub
              type="text"
              value="answer"
            />
          </div>
        </b-list-group-item-stub>
        <b-list-group-item-stub
          activeclass="active"
          class="item border-0"
          event="click"
          exactactiveclass="active"
          routertag="a"
          tag="div"
          target="_self"
        >
          <label>
            <strong>
              2. question
            </strong>
          </label>
           
          <div>
            <b-form-radio-group-stub
              buttonvariant="secondary"
              options="Yes,No"
            />
          </div>
        </b-list-group-item-stub>
        <b-list-group-item-stub
          activeclass="active"
          class="item border-0"
          event="click"
          exactactiveclass="active"
          routertag="a"
          tag="div"
          target="_self"
        >
          <label>
            <strong>
              3. question
            </strong>
          </label>
           
          <div>
            <b-form-select-stub
              options="first,second,third"
              selectsize="0"
              value="first"
            />
          </div>
        </b-list-group-item-stub>
        <b-list-group-item-stub
          activeclass="active"
          class="item border-0"
          event="click"
          exactactiveclass="active"
          routertag="a"
          tag="div"
          target="_self"
        >
          <label>
            <strong>
              4. question
            </strong>
          </label>
           
          <div>
            <datepicker-stub
              format="MM/dd/yyyy"
              inputclass="form-control"
              language="[object Object]"
              maximumview="year"
              minimumview="day"
              placeholder="Enter Date"
              value="1519138199000"
            />
          </div>
        </b-list-group-item-stub>
      </b-list-group-stub>
    </b-card-stub>
  </b-card-group-stub>
   
  <b-row-stub
    class="my-2"
    tag="div"
  >
    <b-col-stub
      tag="div"
    >
      <button-cmp-stub
        class="w-100"
        type="submit"
        variant="primary"
      >
        Save
      </button-cmp-stub>
    </b-col-stub>
     
    <b-col-stub
      tag="div"
    >
      <button-cmp-stub
        variant="link"
      >
        Cancel
      </button-cmp-stub>
    </b-col-stub>
     
    <b-col-stub
      col="true"
      md="auto"
      tag="div"
    />
  </b-row-stub>
   
  <formchangereasonmodal-stub />
</b-form-stub>
`;
